extensions/amp-a4a/0.1/test/test-a4a-integration.js

```diff
@@ -0,0 +1,187 @@
+      requireAmpResponseSourceOrigin: true,
+    }).onFirstCall().returns(Promise.resolve(mockResponse));
+    adConfig['mock'] = {};
+    a4aRegistry['mock'] = () => {return true;};
+    return createIframePromise().then(f => {
+      fixture = f;
+      installDocService(fixture.win, /* isSingleDoc */ true);
+      upgradeOrRegisterElement(fixture.win, 'amp-a4a', MockA4AImpl);
+      const doc = fixture.doc;
+      a4aElement = doc.createElement('amp-a4a');
+      a4aElement.setAttribute('width', 200);
+      a4aElement.setAttribute('height', 50);
+      a4aElement.setAttribute('type', 'mock');
+    });
+  });
+
+  afterEach(() => {
+    sandbox.restore();
+    resetScheduledElementForTesting(window, 'amp-a4a');
+    delete adConfig['mock'];
+    delete a4aRegistry['mock'];
+  });
+
+  it('should render a single AMP ad in a friendly iframe', () => {
+    return fixture.addElement(a4aElement).then(unusedElement => {
+      expectRenderedInFriendlyIframe(a4aElement, 'Hello, world.');
+    });
+  });
+
+  it('should fall back to 3p when no signature is present', () => {
+    mockResponse.headers.delete(SIGNATURE_HEADER);
+    return fixture.addElement(a4aElement).then(unusedElement => {
+      expectRenderedInXDomainIframe(a4aElement, TEST_URL);
+    });
+  });
+
+  it('should fall back to 3p when the XHR fails', () => {
+    xhrMock.resetBehavior();
+    xhrMock.throws(new Error('Testing network error'));
+    // TODO(tdrl) Currently layoutCallback rejects, even though something *is*
+    // rendered.  This should be fixed in a refactor, and we should change this
+    // .catch to a .then.
+    return fixture.addElement(a4aElement).catch(error => {
+      expect(error.message).to.contain.string('Testing network error');
+      expect(error.message).to.contain.string('amp-a4a:');
+      expectRenderedInXDomainIframe(a4aElement, TEST_URL);
+    });
+  });
+
+  it('should fall back to 3p when extractCreative throws', () => {
+    sandbox.stub(MockA4AImpl.prototype, 'extractCreativeAndSignature').throws(
+        new Error('Testing extractCreativeAndSignature error'));
+    // TODO(tdrl) Currently layoutCallback rejects, even though something *is*
+    // rendered.  This should be fixed in a refactor, and we should change this
+    // .catch to a .then.
+    return fixture.addElement(a4aElement).catch(error => {
+      expect(error.message).to.contain.string(
+          'Testing extractCreativeAndSignature error');
+      expect(error.message).to.contain.string('amp-a4a:');
+      expectRenderedInXDomainIframe(a4aElement, TEST_URL);
+    });
+  });
+
+  it('should fall back to 3p when extractCreative returns empty sig', () => {
+    sandbox.stub(MockA4AImpl.prototype, 'extractCreativeAndSignature')
+        .onFirstCall().returns({
+          creative: utf8Encode(validCSSAmp.reserialized),
+          signature: null,
+        })
+        .onSecondCall().throws(new Error(
+            'Testing extractCreativeAndSignature should not occur error'));
```

Comment:
is this to make sure there's no second call to the function?
if so, it might be clearer to do
```
expect(functionStub).to.be.calledOnce;
```